### ISSUE TITLE
5-1 질문드립니다.

### ISSUE THREAD

**Issue by haekyu** [snu-sf-class/PP2015f_TA#93]
안녕하세요 조교님?
case-sum이라는 함수의 타입이 case-sum : (τ → σ) ∗ (τ' → σ) ∗ (τ + τ') → σ 이렇게 되어있는데요,
(case-sum f1 f2 s) 일 때, s의 타입이 τ이면 (f1 s)를 실행시키고, s의 타입이 τ'이면 (f2 s)를 실행시켜라.
라고 제가 이해했습니다.
1. 일단 저렇게 이해한 것이 맞나요?
2. 1이 맞다면, 제가 해야할 것은 s의 타입을 체크해야하는 것인데 어떻게 타입을 체크해야하는지 방향을 전혀 못잡겠습니다ㅠㅠ 혹시 괜찮으시다면 힌트라도 조금 주실 수 있는지요..
3. 그리고 만들기함수인 inl과 inr은 무엇을 만들라고 하는지 감이 잘 안잡히네요. 무엇을 만들라는 것인가요????? case-sum이 사용하기 함수이니까, inl과 inr로 만든 '무엇'인가를 case-sum에서 사용할 수 있게끔 inl과 inr을 짜야하는 것이라고 생각했는데요. 저 '무엇'이 뭘 말하는 건지를 잘 모르겠습니다.

감사합니다.^^

**Comment by hyunminch** [comment]
저도 수강생이지만, tau + tau' 타입을 새로 정의하는 것과 마찬가지인 것 같습니다.

inl는 tau 타입을 받아들여 tau + tau' 타입을 만들고,
inr는 tau' 타입을 받아들여 tau + tau' 타입을 만드는 함수입니다.

tau + tau' 타입을 새로 정의하는 것인 셈이기 때문에, inl은 새로 만들어지는 값이 tau 타입이라는 정보와 tau 값이 무엇인지 명시해주면 되고, inr은 tau'에 대해 마찬가지입니다.

**Comment by haekyu** [comment]
아, 그러면 tau+tau'라는 것이 tau또는 tau'(polymorphic하게?)를 말하는 것이 아니고 아예 새로운 tau+tau'타입(물론 내부 구성은 tau또는 tau'라고 이해할 수는 있지만) 이라는 건가요? 답변 감사합니다!

**Comment by haekyu** [comment]
+추가로 질문드리자면, tau타입이 primitive type만 들어올 수 있다고 생각해도 되나요? hyunmin님이 달아주신 답변을 보면 'inl에서 새로 만들어질 값이 tau 타입이다'를 명시해야 한다고 하셨습니다. 이를 보면 'inl의 인풋의 타입이 tau이다'를 알아내야 할 것 같은데, tau 타입이 무한한 종류의 복잡한 타입(예를 들어 이번 hw5-4나 5-5의 form타입)이 가능해진다면 tau 타입을 어떻게 알아낼 수 있는 건가요?

**Comment by kim-yoonseung** [comment]
네, 일단 1번 질문의 경우 이해하신 대로가 맞고, 위의 @hyunmin-personal 님의 답변도 맞습니다.

tau+tau'라는 새로운 타입을 정의하는 것입니다. (polymorphic 타입이긴 합니다.)

또한 추가 질문에 대해서는, tau는 어떤 타입이라도 올 수 있습니다. 하지만 inl 또는 inr이 인풋을 받아 뭔가를 만들어낼 때 인풋의 타입을 알 필요는 없고, (tau+tau') 에 들어가는 두 가지 타입, 즉 왼쪽 오른쪽 중 어느 타입의 값인지만 표시하면 case-sum 함수에서 사용하는 데 문제가 없을 것입니다.
